Description: Flush the serial port after upload to work around potential driver bugs
Author: Cristian Maglie <c.maglie@bug.st>
--- a/src/EefcFlash.cpp
+++ b/src/EefcFlash.cpp
@@ -62,7 +62,7 @@
       _regs(regs), _canBrownout(canBrownout), _eraseAuto(true)
 {
     assert(planes == 1 || planes == 2);
-    assert(pages <= 1024);
+    assert(pages <= 2048);
     assert(lockRegions <= 32);
 
     // SAM3 Errata (FWS must be 6)
@@ -236,6 +236,8 @@
 {
     waitFSR();
     writeFCR0(enable ? EEFC_FCMD_SGPB : EEFC_FCMD_CGPB, (_canBrownout ? 3 : 1));
+    waitFSR();
+    usleep(10000);
 }
 
 void
--- a/src/PosixSerialPort.cpp
+++ b/src/PosixSerialPort.cpp
@@ -36,7 +36,8 @@
 #endif
 
 PosixSerialPort::PosixSerialPort(const std::string& name, bool isUsb) :
-    SerialPort(name), _devfd(-1), _isUsb(isUsb), _timeout(0)
+    SerialPort(name), _devfd(-1), _isUsb(isUsb), _timeout(0),
+    _autoFlush(false)
 {
 }
 
@@ -237,7 +238,11 @@
     if (_devfd == -1)
         return -1;
 
-    return ::write(_devfd, buffer, len);
+    int res = ::write(_devfd, buffer, len);
+    // Used on macos to avoid upload errors
+    if (_autoFlush)
+        flush();
+    return res;
 }
 
 int
@@ -279,3 +284,8 @@
     return true;
 }
 
+void
+PosixSerialPort::setAutoFlush(bool autoflush)
+{
+    _autoFlush = autoflush;
+}
--- a/src/PosixSerialPort.h
+++ b/src/PosixSerialPort.h
@@ -42,11 +42,13 @@
 
     bool timeout(int millisecs);
     void flush();
+    void setAutoFlush(bool autoflush);
 
 private:
     int _devfd;
     bool _isUsb;
     int _timeout;
+    bool _autoFlush;
 };
 
 #endif // _POSIXSERIALPORT_H
